opengrok-tools/src/main/python/opengrok_tools/utils/mirror.py

```diff
@@ -0,0 +1,400 @@
+                         format(ignored_repos))
+
+        hooks = project_config.get(HOOKS_PROPERTY)
+        if hooks:
+            for hookname in hooks:
+                if hookname == "pre":
+                    prehook = os.path.join(hookdir, hooks['pre'])
+                    logger.debug("pre-hook = {}".format(prehook))
+                elif hookname == "post":
+                    posthook = os.path.join(hookdir, hooks['post'])
+                    logger.debug("post-hook = {}".format(posthook))
+
+        if project_config.get(PROXY_PROPERTY):
+            logger.debug("will use proxy")
+            use_proxy = True
+
+    if not ignored_repos:
+        ignored_repos = []
+
+    return prehook, posthook, hook_timeout, command_timeout, \
+        use_proxy, ignored_repos
+
+
+def mirror_project(config, project_name, check_incoming, uri,
+                   source_root):
+    """
+    Mirror the repositories of single project.
+    :param config global configuration dictionary
+    :param project_name: name of the project
+    :param check_incoming:
+    :param uri
+    :param source_root
+    :return exit code
+    """
+
+    logger = logging.getLogger(__name__)
+
+    project_config = get_project_config(config, project_name)
+    prehook, posthook, hook_timeout, command_timeout, use_proxy, \
+        ignored_repos = get_project_properties(config, project_config,
+                                               config.get(HOOKDIR_PROPERTY))
+
+    proxy = None
+    if use_proxy:
+        proxy = config.get(PROXY_PROPERTY)
+
+    # We want this to be logged to the log file (if any).
+    if project_config:
+        if project_config.get(DISABLED_PROPERTY):
+            logger.info("Project {} disabled, exiting".
+                        format(project_name))
+            return CONTINUE_EXITVAL
+
+    #
+    # Cache the repositories first. This way it will be known that
+    # something is not right, avoiding any needless pre-hook run.
+    #
+    ret = 0
+    repos = get_repos_for_project(project_name,
+                                  ignored_repos,
+                                  commands=config.
+                                  get(COMMANDS_PROPERTY),
+                                  proxy=proxy,
+                                  command_timeout=command_timeout,
+                                  source_root=source_root,
+                                  uri=uri)
+    if not repos:
+        logger.info("No repositories for project {}".
+                    format(project_name))
+        return CONTINUE_EXITVAL
+
+    # Check if any of the repositories contains incoming changes.
+    if check_incoming:
+        got_incoming = False
+        for repo in repos:
+            try:
+                if repo.incoming():
+                    logger.debug('Repository {} has incoming changes'.
+                                 format(repo))
+                    got_incoming = True
```

Comment:
I wonder if it makes sense to reindex only the repository which actually changed (if it's supported at this point by the indexer)

Reply:
No, I meant a general idea for an enhancement. As I see it, now when we get incoming=True from **any** of the repositories (there might by several in one project), we continue:

1) Synchronize all repositories in the project (which is not necessary for these which got incoming=False
2) Run the indexer for the whole project (again not necessary as other repositories have not changed).